libuavcan/include/libuavcan/media/interfaces.hpp

```diff
@@ -0,0 +1,255 @@
+ * on where the media layer objects are placed for a given system or if an implementation chooses to use dynamic
+ * memory internally. Media layer implementations should document this and all other paramters that affect
+ * the amount and type of memory used for a particular system and provide guidance for tuning performance versus
+ * memory size to the user.
+ *
+ * @tparam  FrameT          The media-specific frame type to be exchanged across this interface.
+ * @tparam  MaxTxFrames     The maximum number of frames that can be written in a single operation using
+ *                          this interface. This value must be > 0. If set to 1 then specializations may
+ *                          use different system APIs then if set to > 1.
+ * @tparam  MaxRxFrames     The maximum number of frames that can be read in a single operation using
+ *                          this interface. This value must be > 0. If set to 1 then specializations may
+ *                          use different system APIs then if set to > 1.
+ */
+template <typename FrameT, std::size_t MaxTxFrames = 1, std::size_t MaxRxFrames = 1>
+class LIBUAVCAN_EXPORT Interface
+{
+public:
+    virtual ~Interface() = default;
+
+    static_assert(MaxTxFrames > 0, "MaxTxFrames must be > 0");
+    static_assert(MaxRxFrames > 0, "MaxRxFrames must be > 0");
+
+    /**
+     * The media-specific frame type exchanged across this interface.
+     */
+    using FrameType = FrameT;
+
+    /**
+     * The length of arrays used to write frames through this interface.
+     */
+    static constexpr std::size_t RxFramesLen = MaxRxFrames;
+
+    /**
+     * The length of arrays used to read frames from this interface.
+     */
+    static constexpr std::size_t TxFramesLen = MaxTxFrames;
+
+    /**
+     * Return the index for this interface. The interface index is the canonical identifier used by libuavcan
+     * to open, close, and access a given interface. Per the v1 specification, lower indicies are preferred
+     * when receiving messages through redundant groups.
+     *
+     * @return This interface's index.
+     */
+    virtual std::uint_fast8_t getInterfaceIndex() const = 0;
+
+    /**
+     * Non-blocking transmission. All implementations will have some intermediate buffer this method
+     * writes to since it does not block on actual transmission of the frame. For some implementations,
+     * this method may borrow CPU time to move another, higher priority frame into a lower-level queue
+     * after enqueueing the given one in an intermediate queue.
+     *
+     * @note Implementations are allowed to provide queues based on message priority. Because of this,
+     * if a given message cannot be written the media layer should keep trying to write other messages
+     * with a different priority.
+     *
+     * @param  frames       1..* frames to write into the system queues for immediate transmission.
```

Comment:
This interface doesn't require that. It's up to the caller of the interface to determine what and how many frames to actually transmit. This is a dumb "do it" interface.

Reply:
Sure. My question was rather, can the implementation expect that all frames in a given call will share the same CAN ID? This may enable some optimizations, such as avoiding the need to check priority inversion for every frame in the write group. This is not hugely important but if you foresee that all frames in the group will belong to same transfer always (therefore, will share the same CAN ID) then it should be stated here.

Reply:
There is no such guarantee. Priority inversion concerns must happen above and below this interface (i.e. we'll address this in the transport layer and linux integrators should address this in the socketCAN layer using an appropriate qdisc). My design here is specifically focused on a minimal definition of required hardware/platform support for a given bus and the ability to independently operate a media layer implementation without needing the upper layers of libuavcan. Optimizations should be facilitated (i.e. why I added the vectorized read and write  APIs) but  not required  by this design. If I guarantee the properties of the contents of these vectors I'll encourage complexity in the media layer where it doesn't belong.

Our assumption should be (and perhaps I should add this to the documentation) that calling write will provide a very direct path to the hardware buffers in the peripheral especially on embedded systems where these writes might actually go directly into DMA. This means that our transports will have determined that the set of messages provided are the best to be sent as soon as possible. Cancellation should happen above this call since hardware support for message cancellation is mixed. Perhaps in the future we'll decide to add a different API appropriate for AUTOSAR compatible CAN peripherals which are required to support tx cancellation.

Reply:
Okay, thank you for the elaboration.